samtranslator/validator/validator.py

```diff
@@ -1,55 +1,262 @@
+    # We remove them to be consistent between python versions
+    # Example: "u'integer'" -> "'integer'"
+    UNICODE_TYPE_REGEX = re.compile("u('[^']+')")
+
+    def __init__(self, schema=None):
         """
-        Is this a valid SAM template dictionary
+        Constructor
 
-        :param dict template_dict: Data to be validated
-        :param dict schema: Optional, dictionary containing JSON Schema representing SAM template
-        :return: Empty string if there are no validation errors in template
+        Parameters
+        ----------
+        schema_path : str, optional
+            Path to a schema to use for validation, by default None, the default schema.json will be used
         """
+        # Replaces the static method with the instance method
+        # Otherwise the instance would call the static one
+        # Static one is kept for backward compatibility
+        self.validate = self._validate
 
         if not schema:
-            schema = SamTemplateValidator._read_schema()
+            schema = self._read_json(sam_schema.SCHEMA_NEW_FILE)
 
-        validation_errors = ""
+        # Helps resolve the $Ref to external files
+        # schema content must be passed to resolve local (#/def...) references
+        # Hardcoded "/" as this is an URI, not a Path
+        resolver = jsonschema.RefResolver("file://" + sam_schema.SCHEMA_DIR + "/", schema)
 
-        try:
-            jsonschema.validate(template_dict, schema)
-        except ValidationError as ex:
-            # Stringifying the exception will give us useful error message
-            validation_errors = str(ex)
-            # Swallowing expected exception here as our caller is expecting validation errors and
-            # not the valiation exception itself
-            pass
-
-        return validation_errors
+        SAMValidator = jsonschema.validators.extend(
+            jsonschema.Draft7Validator,
+            type_checker=jsonschema.Draft7Validator.TYPE_CHECKER.redefine_many(
+                {"object": is_object, "intrinsic": is_intrinsic}
+            ),
+        )
+        self.validator = SAMValidator(schema, resolver=resolver)
 
     @staticmethod
-    def _read_schema():
+    def validate(template_dict, schema=None):  # pylint: disable=method-hidden
+        """
+        Validates a SAM Template
+
+        [DEPRECATED]: Instanciate this class and use the instance method instead:
+            validator = SamTemplateValidator()
+            validator.validate(template_dict)
+
+        Parameters
+        ----------
+        template_dict : dict
+            Template
+        schema : dict, optional
+            Schema content, by default None
+
+        Returns
+        -------
+        list[str]
+            List of validation errors
         """
-        Reads the JSON Schema at given file path
+        validator = SamTemplateValidator(schema)
+
+        return ", ".join(validator.validate(template_dict))
```

Comment:
Should this be `validator._validate(template_dict)`? I don't see a `validate` method that isn't static here. If there is and I just don't see it (due to not enough coffee this morning), I would suggest making the static and instance method different names to help the reader not to think this is an infinite loop in calling itself.

Reply:
Calling a private function on an instance in generally not a good practice and frowned upon by pylint.
`validate` is overridden in the constructor with a call to the private one.
We kept the static one for backward compatibility, added a Deprecated notice in the doc.
Not sure about a good new name for the instance validate, if we want to change it.

Reply:
Renamed the instance validate to `get_errors` to remove the ambiguity.